### DISCUSSION TITLE
tensordot and einsum | integrating with other libraries

### DISCUSSION THREAD

**Discussion by ShHsLin** [NVIDIA/cuQuantum#111] [Q&A]
Hi,

I am exploring the possibility of using cuQuantum to speed up standard classical simulations of quantum many-body systems. One use case will be the simulation of PEPS, where the tensor is of rank (order) 5 and higher and contraction is complicated. As a user, I prefer to stick coding with python libraries, e.g. JAX, pytorch, etc... and I hope to still get the benefit of speedup with cuQuantum. This leads to several different questions that I would appreciate if I could get some pointers here.

I assume the followings, please correct me if I am wrong.

1. tensordot and eisnum are the bottleneck of my computation.
2. JAX, pytorch have their own implementation of tensordot, einsum and are not using cuQuantum functions.
3. tensordot and einsum implementation of cuQuantum are faster  

I ask the benchmark on this in the separate thread [here](https://github.com/NVIDIA/cuQuantum/discussions/110) and assume it to be true here.

I wonder as a user how hard it is for me to change and link the tensordot, einsum of a given library to use cuQuantum's function or has anyone already tried to do this.

**Comment by DmitryLyakh** [comment]
The cuTensorNet module of the cuQuantum library provides Python API for performing a contraction of an arbitrary tensor network as well as computing the gradients of the tensor network contraction via back-propagation that is compatible with PyTorch. You can find details in the cuQuantum Python documentation (https://docs.nvidia.com/cuda/cuquantum/22.07.1/python/index.html). @yangcal @leofang for viz
